sklearn/model_selection/_search.py

```diff
@@ -576,6 +581,37 @@ def classes_(self):
         self._check_is_fitted("classes_")
         return self.best_estimator_.classes_
 
+    @abstractmethod
+    def _generate_candidates(self):
+        """Supplies candidates to search in response to accumulated results
+
+        To be overridden by implementers.
+
+        As in the following snippet, an implementation should yield an
+        initial list of candidates, where each candidate is a dict of
+        parameter settings.  The yield expression then returns the results
+        dict corresponding to those candidates, and further yields can
+        be used to evaluate more candidates.
+
+        ::
+
+            def _generate_candidates(self):
+                results = yield self.initial_candidates()
+                while self.not_good_enough(results):
+                    results = yield self.more_candidates()
+
```

Comment:
Could we refactor this function? I think a better implementation would be

``` python
def _generate_candidates(self):
    params = results = None
    while True:
        params = self._candidates(params, results)
        if params is None:
            break
        results = yield params
```

where `self._candidates` should be overwritten (and raise a `NotImplementedError` in this class).

Reply:
This would allow the documentation of `_candidates` to be simple: takes in params and results, and returns more params. Plus, the iteration logic is separated from the choosing logic.